async-local-logger.js

```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+const Http = require('http');
+const { AsyncLocalStorage } = require('async_hooks');
+const Toys = require('toys');
+
+const internals = {};
+
+exports.log = (...args) => {
+
+    const { asyncLocal, storeMap } = internals;
+    const request = storeMap.get(asyncLocal.getStore());
+
+    if (request) {
+        request.log(...args);
+    }
+};
+
+exports.plugin = {
+    name: 'async-local-logger',
+    register(srv) {
+
+        srv.ext(Toys.onRequest((request, h) => {
+
+            const { req } = request.raw;
+            const { storeMap } = internals;
+
+            storeMap.set(req, request);
+
+            return h.continue;
+        }));
+    }
+};
+
+exports.createHttpServer = (...args) => {
+
+    const srv = Http.createServer(...args);
+    const { asyncLocal } = internals;
+
+    (async () => {
+
+        let listener;
+
+        while (!listener) {
+
+            const [event, maybeListener] = await Toys.event(srv, 'newListener', {
+                error: false,
+                multiple: true
+            });
+
+            if (event === 'request') {
```

Comment:
So does this while loop spin until the first request gets in? And from then on it swaps the listener on each request using those srv `request` listeners?

Reply:
Oh it doesn't _really_ _spin_ cuz it's waiting for events right there gotcha

Reply:
It waits until hapi registers its `'request'` listener, then swaps that listener out for the same listener, but wrapped in an async-local context.

Reply:
(And that happens exactly once, not per request.)

Reply:
Oh gotcha! I was assuming `srv.on('request'` acted like our hapi lifecycle hook

Reply:
Oh wait does it? I'm confused — is `asyncLocal.run` down below being run for every request?